### PR TITLE
Resolves #991: Perform record fetches after union/intersection.

### LINKED ISSUE #991: Perform record fetches after union/intersection operators where possible
Generated plans do not minimize the number of expensive index lookups.

 Union Optimization

 *                                  Starting Plan
 *
 *                                      UnionPlan
 *                                       /              \
 *                                     /                \
 * RecordQueryPlan (IndexFetch) RecordQueryPlan (IndexFetch)
 *
 *                                          |
 *                                          |
 *                                          V
 *
 *                                  Transformed Plan
 *
 *                                  UnionPlan (IndexFetch)
 *                                  /              \
 *                                 /                \
 *                RecordQueryPlan   RecordQueryPlan

Intersection Optimization

 *                                  Starting Plan
 *
 *                                  IntersectionPlan
 *                                  /              \
 *                                 /                \
 *                 RecordQueryPlan (IndexFetch) RecordQueryPlan (IndexFetch)
 *
 *                                          |
 *                                          |
 *                                          V
 *
 *                                  Transformed Plan
 *
 *                                  IntersectionPlan (IndexFetch)
 *                                  /              \
 *                                 /                \
 *                         RecordQueryPlan   RecordQueryPlan

UnorderdPrimaryKeyDistinct

 *                               UnorderedPrimaryKeyDistinct
 *                                          |
 *                                          |
 *                             RecordQueryPlanWithIndex (Index Lookup)
 *
 *                                         ||
 *                                         ||
 *                                         V
 *
 *                               UnorderedPrimaryKeyDistinct (Index Lookup)
 *                                          |
 *                                          |
 *                             RecordQueryPlanWithIndex (Covering Index)
 *
 *
￼

### PR DESCRIPTION
**Pull request by nschiefer**
Implements a delicate-but-working pushdown of union, intersection, and primary key distinct operations below the record fetch (from an index scan) within the `RecordQueryPlanner`.

This is a fairly delicate operation that isn't easy to express properly within the `RecordQueryPlanner`. In a number of places, the transformation as expressed is incorrect, but happens to be correct in all the circumstances where we happen to use it.

Because the execution framework cannot currently manipulate streams of objects other than records, index entries are represented as partial records, as in covering index scans (credit to @jleach4 for this idea).

This is based on @jleach4's work on jleach4/991.

### CODE REVIEW

**Comment by nschiefer** [comment]
This still needs some more testing but that shouldn't block the beginning of a review.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java [hunk 1]
```diff
@@ -230,6 +231,38 @@ public static FanType valueOf(RecordMetaDataProto.Field.FanType fanType) throws
     @Nonnull
     KeyExpression normalizeForPlanner(@Nonnull Source source, @Nonnull List<String> fieldNamePrefix);
 
+    /**
+     * Return the key fields for an expression.
+     * @param rootExpression a key expression
+     * @return the parts of the key expression that are in the key
+     */
+    @API(API.Status.EXPERIMENTAL)
+    static List<KeyExpression> getKeyFields(@Nonnull KeyExpression rootExpression) {
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java:240]: John had it that way but I think this is preferable. The awkward thing about the key fields/value fields distinction is that it's really only conceptually defined on the root expression (which might be a `KeyWithValueExpression`).

I have an old issue somewhere about splitting the `KeyExpression` interface into a `KeyExpression` (which is a true tree-like data structure) and `IndexExpression` (which might be a `KeyWithValueExpression`, `GroupingExpression`, or  just a plain `KeyExpression`), but obviously that's a highly invasive change.

Given that, I'd rather keep these as static methods, since that seems to contain the "conceptual bleed". I'm willing to change it up if there's strong disagreement, though.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java [hunk 2]
```diff
@@ -230,6 +231,38 @@ public static FanType valueOf(RecordMetaDataProto.Field.FanType fanType) throws
...
+        }
+    }
+
+    /**
+     * Return the value fields for an expression.
+     * @param rootExpression a key expression
+     * @return the parts of the key expression that are in the value
+     */
+    @API(API.Status.EXPERIMENTAL)
+    static List<KeyExpression> getValueFields(@Nonnull KeyExpression rootExpression) {
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java:256]: See above.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryCoveringIndexPlan.java [hunk 3]
```diff
@@ -156,9 +156,6 @@ public boolean equals(Object o) {
         if (o == null || getClass() != o.getClass()) {
             return false;
         }
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryCoveringIndexPlan.java:158]: Yes, we'll definitely  need some careful rebasing work.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryFetchFromPartialRecordPlan.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    @Override
+    public String toString() {
+        return "Fetch(" + getChild().toString() + ")";
+    }
+
+    @Nonnull
+    @Override
+    public PlannerGraph rewritePlannerGraph(@Nonnull final List<? extends PlannerGraph> childGraphs) {
+        return PlannerGraph.fromNodeAndChildGraphs(
+                new PlannerGraph.OperatorNodeWithInfo(this, NodeInfo.LOAD_BY_KEYS_OPERATOR),
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryFetchFromPartialRecordPlan.java:132]: Do you have a `NodeInfo` that you'd rather use, or should I just create a new one?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryPlan.java [hunk 5]
```diff
@@ -117,4 +121,29 @@
...
+        for (Quantifier childQuantifier : getQuantifiers()) {
+            if (!(childQuantifier instanceof Quantifier.Physical)) {
+                throw new RecordCoreException("quantifiers of RecordQueryPlans must be physical");
+            }
+            // Group expression is the only type of reference at this point.
+            GroupExpressionRef<RecordQueryPlan> childGroup = ((GroupExpressionRef)((Quantifier.Physical)childQuantifier).getRangesOver());
+            RecordQueryPlan child = childGroup.get(); // Group is generated by the RecordQueryPlanner so must have a single member
+
+            RecordQueryPlan modifiedChild = child.accept(visitor);
+            if (child != modifiedChild) { // intentional use of reference equality, since equals() might not be conservative enough for plans
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryPlan.java:142]: I'm happy to make this nullable and return null instead of an unchanged plan if the visitor doesn't make any changes. Is that preferable?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    @Nullable
+    private final KeyExpression commonPrimaryKey;
+
+    public RecordQueryPlannerSubstitutionVisitor(@Nonnull RecordMetaData recordMetadata, @Nullable KeyExpression commonPrimaryKey) {
+        this.recordMetadata = recordMetadata;
+        this.commonPrimaryKey = commonPrimaryKey;
+    }
+
+    public static RecordQueryPlan applyVisitors(@Nonnull RecordQueryPlan plan, @Nonnull RecordMetaData recordMetaData, @Nullable KeyExpression commonPrimaryKey) {
+        return plan.accept(new UnorderedPrimaryKeyDistinctVisitor(recordMetaData, commonPrimaryKey))
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:61]: Yes, we walk it four times. I don't think that's a big deal, and it's really hard to improve on without basically recreating the memo data structure from the Cascades planner. 🙃

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                .accept(new UnionVisitor(recordMetaData, commonPrimaryKey))
+                .accept(new IntersectionVisitor(recordMetaData, commonPrimaryKey))
+                .accept(new UnorderedPrimaryKeyDistinctVisitor(recordMetaData, commonPrimaryKey));
+    }
+
+    @Nonnull
+    public abstract RecordQueryPlan visit(@Nonnull RecordQueryPlan recordQueryPlan);
+
+    @Nullable
+    protected RecordQueryPlan removeIndexFetch(@Nonnull RecordQueryPlan plan) {
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:71]: Originally, the visitor was an interface and there was a `BaseMetadataVisitor` as a common super class of all the visitors we have. It seemed sort of strange to have an interface with exactly one implementation (which was an abstract class of all things).

I think I agree in general, but in this particular case I don't think we're trying to build much of a "conceptual model" with the visitors here (unlike with the rules in the new planner). I don't mind separating the interface and the abstract base class, though.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        if (plan instanceof RecordQueryPlanWithIndex) {
+            RecordQueryPlanWithIndex indexPlan = (RecordQueryPlanWithIndex) plan;
+            Index index = recordMetadata.getIndex(indexPlan.getIndexName());
+            final Collection<RecordType> recordTypes = recordMetadata.recordTypesForIndex(index);
+            final IndexKeyValueToPartialRecord keyValueToPartialRecord = generateIndexKeyValueToPartialRecord(indexPlan);
+            if (keyValueToPartialRecord != null && recordTypes.size() == 1) {
+                RecordType recordType = recordTypes.iterator().next();
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:78]: Changed.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                return new RecordQueryCoveringIndexPlan(indexPlan, recordType.getName(), keyValueToPartialRecord);
```
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:79]: Thanks :) Getting this all lined up just-so was tricky but I'm happy with the result.

**Comment by nschiefer** [comment]
I'm working on a full-text test, but the infrastructure around those test cases is wonky in a way that's making it tricking to build the appropriate `ParameterizedTest`s. I also have some tests that probe the real edge cases of the transformations.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/KeyExpression.java:240]: Fine either way.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryPlan.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryPlan.java:142]: No big deal, it is doing the same thing. It's just style. In some visitor pattern variations you can even reject the `accept` or the `visit()`. In other variations you return `Optional<...>` with a default implementation returning `Optional.empty()`, so you don't have to overwrite it for every subclass. In any case it's producing something, it just matters it is correctly interpreted.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:61]: Well, you could dispatch within the visitor but I agree that makes the code harder to understand and it undermines orthogonality. I also agree with the statement that it shouldn't be a big deal.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlannerConfiguration.java [hunk 10]
```diff
@@ -35,15 +35,18 @@
...
     private final int complexityThreshold;
     private boolean checkForDuplicateConditions;
+    private boolean pushFetchAboveUnionAndIntersection;
 
     private RecordQueryPlannerConfiguration(@Nonnull QueryPlanner.IndexScanPreference indexScanPreference,
                                             boolean attemptFailedInJoinAsOr,
                                             int complexityThreshold,
-                                            boolean checkForDuplicateConditions) {
+                                            boolean checkForDuplicateConditions,
+                                            boolean pushFetchAboveUnionAndIntersection) {
```
[the remark this answers is not in the document]
Comment (jleach4) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlannerConfiguration.java:44]: agreed.

[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlannerConfiguration.java:44]: Fixed.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryFetchFromPartialRecordPlan.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryFetchFromPartialRecordPlan.java:132]: Created `NodeInfo.FETCH_OPERATOR`.

**Comment by nschiefer** [comment]
The Sonarqube job failed with a really strange error about a missing cluster file from one of the tests.

test this please

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    public abstract RecordQueryPlan visit(@Nonnull RecordQueryPlan recordQueryPlan);
+
+    @Nullable
+    protected RecordQueryPlan removeIndexFetch(@Nonnull RecordQueryPlan plan) {
+        if (plan instanceof RecordQueryPlanWithIndex) {
+            RecordQueryPlanWithIndex indexPlan = (RecordQueryPlanWithIndex) plan;
+            Index index = recordMetadata.getIndex(indexPlan.getIndexName());
+            final Collection<RecordType> recordTypes = recordMetadata.recordTypesForIndex(index);
+            final IndexKeyValueToPartialRecord keyValueToPartialRecord = generateIndexKeyValueToPartialRecord(indexPlan);
+            if (keyValueToPartialRecord != null && recordTypes.size() == 1) {
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:78]: There is duplication between here and `generateIndexKeyValueToPartialRecord`:
* get index from metadata
* get record types from index
* assert only one record type

Since the method is private, it seems like a little bit of refactoring could avoid that.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/match/FetchMatcher.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+package com.apple.foundationdb.record.query.plan.match;
+
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryFetchFromPartialRecordPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+
+import javax.annotation.Nonnull;
+
+public class FetchMatcher extends PlanMatcherWithChild {
```
Comment (MMcM) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/match/FetchMatcher.java:30]: Needs java doc.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/indexes/TextIndexTestUtils.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+import com.apple.foundationdb.record.TestRecordsTextProto;
+import com.apple.foundationdb.record.provider.common.TransformedRecordSerializer;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class TextIndexTestUtils {
```
Comment (MMcM) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/indexes/TextIndexTestUtils.java:32]: Needs java doc.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java (hunk 11 above)
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:78]: Extracted into a private method.

Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:78]: Did this get pushed?
In addition to the source code duplication, can't we avoid doing these more than one in quick succession?

Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/visitor/RecordQueryPlannerSubstitutionVisitor.java:78]: No, not yet. I'm still getting the comparison key validation logic working. I think I've found a way to pull this out into a single spot per evaluation.

**Comment by nschiefer** [comment]
I was able to unify the covering index and delayed fetch optimization flows so that checking for the fields in the comparison key of an ordered union/intersection uses the same logic as the filters in Mike's PR (#1001).

**Comment by nschiefer** [comment]
Known test failure unrelated to this change:
```
com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerBuildValueIndexTest$Unsafe &gt; fiftySplitElements() FAILED
    com.apple.foundationdb.record.provider.foundationdb.FDBExceptions$FDBStoreRetriableException: Request for future version

        Caused by:
        com.apple.foundationdb.FDBException: Request for future version
```

test this please

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java [hunk 14]
```diff
@@ -1438,52 +1439,62 @@ private RecordQueryPlan tryToConvertToCoveringPlan(@Nonnull PlanContext context,
...
+        for (KeyExpression field: fields) {
+            if (!addCoveringField(field, builder, keyFields, valueFields)) {
+                return null;
             }
         }
 
+        // Validity check ensures that we don't attempt this transformation if there are any repeated fields..
         if (!builder.isValid()) {
-            return chosenPlan;
+            return null;
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java:1492]: `chosenPlan` ?

Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java:1492]: oh got it -- different contract in a different method

Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java:1492]: `chosenPlan` is the index plan before the covering index transformation. I pulled out the building logic from the convert-to-covering logic so that it could be reused by the visitor. Returning `null` is the equivalent for the new control flow.
